### PR TITLE
Update docs on random seed

### PR DESCRIPTION
**Pull request by hanlint**
Our current docstrings on seeding were incorrect, leading to an issue reported in #793 . 

When using `seed_all` convenience functions, the seed must still be passed to the Trainer:
```
trainer = Trainer(
   ...,
   seed=42,
)
```

otherwise, the trainer will choose a seed based on the system time.

This PR updates the docstrings and adds a reproducibility section to the trainer guide.

### CODE REVIEW

File: docs/source/trainer/using_the_trainer.rst [hunk 1]
```diff
@@ -443,6 +443,45 @@ points during training and (2) load them back to resume training later.
...
+
+    trainer = Trainer(
+        ...,
+        seed=42,
+    )
+
+If no seed is provided, a random seed will be generated from system time.
+
+Since the model and dataloaders are initialized outside of the Trainer, for complete
+determinism, we recommend calling :meth:`~.reproducibility.seed_all` and/or
```
Comment (ravi-mosaicml) [inline docs/source/trainer/using_the_trainer.rst:463]: These are not autolinking...see https://mosaicml-composer--794.com.readthedocs.build/en/794/trainer/using_the_trainer.html#reproducibility

**Review by ravi-mosaicml — COMMENTED** [review]
LGTM, thanks for clarifying, though please see comments re: autolinking and ..testcode.

**Review by ravi-mosaicml — APPROVED** [review]
LGTM -- I think sphinx prefers `func` instead of `meth` for unbound functions -- might be why it wasn't auto-linking originally in some places.
